pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,331 @@
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/pkg/log"
+)
+
+var (
+	notReadyEndpointkey                        = "networking.alpha.istio.io/notReadyEndpoints"
+	_                   model.Controller       = &MCPDiscovery{}
+	_                   model.ServiceDiscovery = &MCPDiscovery{}
+)
+
+// DiscoveryOptions stores the configurable attributes of a Control
+type DiscoveryOptions struct {
+	XDSUpdater   model.XDSUpdater
+	Env          *model.Environment
+	ClusterID    string
+	DomainSuffix string
+}
+
+// mixerEnabled checks to see if mixer is enabled in the environment
+// so we can set the UID on eds endpoints
+func (o *DiscoveryOptions) mixerEnabled() bool {
+	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
+}
+
+// MCPDiscovery provides storage for NotReadyEndpoints
+type MCPDiscovery struct {
+	// [ip:port]config
+	notReadyEndpoints   map[string]*model.Config
+	notReadyEndpointsMu sync.RWMutex
+	options             *DiscoveryOptions
+}
+
+// NewMCPDiscovery provides a new instance of Discovery
+func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		options:           options,
+	}
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+// GetProxyServiceInstances returns service instances co-located with a given proxy
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
+		notReadyIP, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		notReadyPort, err := strconv.Atoi(port)
+		if proxyIP != notReadyIP || err != nil {
+			continue
+		}
+		se, ok := conf.Spec.(*networking.ServiceEntry)
+		if !ok {
+			return nil, errors.New("getProxyServiceInstances: wrong type")
+		}
+
+		for _, h := range se.Hosts {
+			for _, svcPort := range se.Ports {
+				// add not ready endpoint
+				out = append(out, d.notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
+				// add other endpoints
+				out = append(out, d.serviceInstances(se, svcPort, conf, h)...)
+			}
+		}
+	}
+	return out, nil
```

Comment:
won't this be empty if all endpoints are "Ready"? since we only iterate over notReadyEndpoints

Reply:
My question is why not only unready endpoints, what about the ready ones

Reply:
Yes, it would be and that is by design. The plan was/is to only expose notReadyEndpoints from this discovery, all the ready endpoints at this point have already been updated either via configUpdate or EDSUpdate throught the controller.